cupyx/scipy/ndimage/filters.py

```diff
@@ -107,9 +107,17 @@ def _correlate_or_convolve(input, weights, output, mode, cval, origin,
         return output
     input = cupy.ascontiguousarray(input)
     weights = cupy.ascontiguousarray(weights, cupy.float64)
+
+    # The kernel needs only the non-zero kernel values and their coordinates.
+    # This allows us to use a single for loop to compute the ndim convolution.
+    # The loop will be over only the the non-zero entries of the filter.
+    wlocs = cupy.where(weights)
+    wvals = weights[wlocs]     # (nnz,) array of non-zero values
+    wlocs = cupy.stack(wlocs)  # (ndim, nnz) array of indices for these values
+
     return _get_correlete_kernel(
-        input.ndim, mode, cval, input.shape, tuple(wshape), tuple(origin))(
-        input, weights, output)
+        input.ndim, mode, cval, input.shape, weights.shape, wvals.size,
```

Comment:
Another point is that, by using `wvals.size` (which can vary depending on the values of `weights`) in the kernel code, it may cause infinitely many instances of kernel compilation.

Reply:
Yes, but wasn't that already the case before as well since the kernel can have any number of different shapes too? (The kernel dimensions get hardcoded into the kernel)

Reply:
I would guess dense kernels are quitr common but there are some interesting cases with sparse ones. For example, all nd filters in [this publication have only two non-zero entries](https://www.google.com/url?sa=t&source=web&rct=j&url=https://arxiv.org/abs/1708.08421&ved=2ahUKEwiI3qCpuMPmAhXMLs0KHTebD3IQFjABegQIBBAF&usg=AOvVaw1H6xZa4nowi3rLNgIH50YU)

I'm not sure there is a real world scenario where thousands of different kernels would be generated.

Reply:
If `weights` is guaranteed to have at most 2 non-zero entries, `convolve` is too generic to calculate it. It seems to be able to be achieved by just only adding two slices.

Reply:
The two-entry case was an extreme example of sparsity, but there are some other intermediate cases. Another one that I have seen are diamond shaped filters (i.e. extent 3 x 3 but the 4 corner entries are zero or 5 x 5 with 13 central non-zero entries). Those cases would also benefit from this PR.